### PR TITLE
7260cx3 DualToR config.bcm support based on DualToR setting in device metadata at boot time

### PR DESCRIPTION
**Pull request by gechiang**
#### Why I did it
In order to support DualToR IPinIP functionality for 7260CX3 SKUs there are new BRCM SOC properties required.  We do not want these new SOC properties to be applied for all 7260CX3 SKUs but to base on the DUT role to decide which DUT should include these new SOC properties at boot time.
 
#### How I did it
Added code to check the device metadata to inspect if the DUT is configured for DualToR functionality. Renamed/Updated the current config.bcm file of the supported 7260cx3 SKUs to optionally include the IPinIP required SOC properties based on the DUT DualToR role.  This file is used for "sonic-cfggen" at boot time to dynamically generate the required config.bcm file to be used by BRCM SDK.  This new method eliminated the need to have multiple versions of config.bcm files where the delta is small.

#### NOTE: 
This PR has pulled in the changes made by (https://github.com/Azure/sonic-buildimage/pull/7027)
and converted the config.bcm selection to use this new method since the MMU selection SOC property is the only line of difference between the T0 and T1 config.bcm files based on the switch role being T0 or T1.

#### How to verify it
Manually changed the device metadata to either include or don't include the DualToR role and verify that the correct bcm.config is installed in the ASIC. 
Here is the verification output for the case where the 7260cx3 is not configured for DualToR purpose:
```
admin@str2-7260cx3-acs-9:~$ docker exec -it syncd bash
root@str2-7260cx3-acs-9:/# cd /etc/sai.d/
root@str2-7260cx3-acs-9:/etc/sai.d# ls -l
total 40
-rw-r--r-- 1 root root 36437 Mar 26 05:20 sai.config
-rw-r--r-- 1 root root    67 Mar 26 05:20 sai.profile
root@str2-7260cx3-acs-9:/etc/sai.d# cat sai.profile
SAI_INIT_CONFIG_FILE=/etc/sai.d/sai.config
SAI_NUM_ECMP_MEMBERS=64
root@str2-7260cx3-acs-9:/etc/sai.d# cat sai.config

PHY_AN_ALLOW_PLL_CHANGE=1
arl_clean_timeout_usec=15000000
asf_mem_profile=2
bcm_num_cos=10
bcm_stat_flags=1
bcm_stat_jumbo=9236
cdma_timeout_usec=15000000
core_clock_frequency=1700
dma_desc_timeout_usec=15000000
dpp_clock_ratio=2:3
higig2_hdr_mode=1
ipv6_lpm_128b_enable=1
l3_alpm_enable=2
lpm_scaling_enable=0
l2xmsg_mode=1
max_vp_lags=0
mem_scan_enable=1
miim_intr_enable=0
module_64ports=1
os=unix
oversubscribe_mode=1
pbmp_xport_xe=0xfffffffd3fffffff4fffffffc7ffffffe
#pbmp_xport_xe=0x47fffffff1fffffffcfffffffe7ffffffe
phy_an_allow_pll_change_hg=0
phy_an_c73=1
phy_chain_rx_lane_map_physical{1.0}=0x3210
phy_chain_rx_lane_map_physical{101.0}=0x0123
...
```
Here is the validation for the case where a 7260CX3 do have DualToR configure:
```
admin@str2-7260cx3-acs-9:~$ bcmcmd "config show" | grep sai_
    sai_tunnel_support=1
admin@str2-7260cx3-acs-9:~$ bcmcmd "config show" | grep ecmp
    l3_ecmp_levels=2
admin@str2-7260cx3-acs-9:~$ bcmcmd "config show" | grep host_as_r
    host_as_route_disable=1

admin@str2-7260cx3-acs-9:~$ docker exec -it syncd bash
root@str2-7260cx3-acs-9:/# cd /etc/sai.d/
root@str2-7260cx3-acs-9:/etc/sai.d# ls -l
total 40
-rw-r--r-- 1 root root 36534 Mar 26 05:32 sai.config
-rw-r--r-- 1 root root    67 Mar 26 05:32 sai.profile
root@str2-7260cx3-acs-9:/etc/sai.d# cat sai.profile
SAI_INIT_CONFIG_FILE=/etc/sai.d/sai.config
SAI_NUM_ECMP_MEMBERS=64
root@str2-7260cx3-acs-9:/etc/sai.d# cat sai.config
sai_tunnel_support=1
                               host_as_route_disable=1
                               l3_ecmp_levels=2
mmu_init_config="MSFT-TH2-Tierr0"
PHY_AN_ALLOW_PLL_CHANGE=1
arl_clean_timeout_usec=15000000
asf_mem_profile=2
bcm_num_cos=10
bcm_stat_flags=1
bcm_stat_jumbo=9236
cdma_timeout_usec=15000000
core_clock_frequency=1700
dma_desc_timeout_usec=15000000
dpp_clock_ratio=2:3
higig2_hdr_mode=1
ipv6_lpm_128b_enable=1
l3_alpm_enable=2
lpm_scaling_enable=0
l2xmsg_mode=1
max_vp_lags=0
mem_scan_enable=1
miim_intr_enable=0
module_64ports=1
os=unix
oversubscribe_mode=1
pbmp_xport_xe=0xfffffffd3fffffff4fffffffc7ffffffe
#pbmp_xport_xe=0x47fffffff1fffffffcfffffffe7ffffffe
phy_an_allow_pll_change_hg=0
phy_an_c73=1
phy_chain_rx_lane_map_physical{1.0}=0x3210
phy_chain_rx_lane_map_physical{101.0}=0x0123
phy_chain_rx_lane_map_physical{105.0}=0x0123
phy_chain_rx_lane_map_physical{109.0}=0x0123
...
```

Here is the result when the same code is running on a non 7260cx3 SKU where the config .bcm setup remains unchanged and continue to load from the original config.bcm file without going through "sonic-cfggen":
```
admin@str-s6000-acs-8:~$ docker exec -it syncd bash
root@str-s6000-acs-8:/# cd /etc/sai.d/
root@str-s6000-acs-8:/etc/sai.d# ls -l
total 4
-rw-r--r-- 1 root root 96 Mar 26 05:22 sai.profile
root@str-s6000-acs-8:/etc/sai.d# cat sai.profile
SAI_INIT_CONFIG_FILE=/usr/share/sonic/hwsku/td2-s6000-32x40G.config.bcm
SAI_NUM_ECMP_MEMBERS=32
root@str-s6000-acs-8:/etc/sai.d#
```

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [x] 202012

#### Description for the changelog
7260cx3 dynamic generate config.bcm based on DualToR role at boot time to add IPinIP support.

### CODE REVIEW

File: device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-Q64/append.config.j2
Comment (lguohan) [inline device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-Q64/append.config.j2:11]: the file name can be config.bcm.j2

File: device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/append.config.j2
Comment (lguohan) [inline device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/append.config.j2:22]: the file name can be config.bcm.j2

File: device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/append.config.j2
Comment (lguohan) [inline device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/append.config.j2:2]: >ock = 'mmu_init_config="MSFT-TH2-Tierr1"' -%} [](start = 13, length = 45)

mmu_init_config="MSFT-TH2-Tier1"

youi have type here.

File: device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/append.config.j2
Comment (lguohan) [inline device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/append.config.j2:8]: check typo

File: device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-D108C8/append.config.j2
Comment (lguohan) [inline device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-D108C8/append.config.j2:9]: prefer to move to the end

File: device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/append.config.j2
Comment (lguohan) [inline device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/append.config.j2:20]: prefer to move to the end

File: platform/broadcom/docker-syncd-brcm/start.sh
Comment (lguohan) [inline platform/broadcom/docker-syncd-brcm/start.sh:24]: >append.config.j2 [](start = 23, length = 16)

config.bcm.j2

File: platform/broadcom/docker-syncd-brcm/start.sh
Comment (lguohan) [inline platform/broadcom/docker-syncd-brcm/start.sh:25]: >sai.config [](start = 68, length = 10)

rename to config.bcm

Comment (lguohan) [inline platform/broadcom/docker-syncd-brcm/start.sh:25]: >append.config.j2  [](start = 38, length = 17)

rename to config.bcm

File: platform/broadcom/docker-syncd-brcm/start.sh
Comment (lguohan) [inline platform/broadcom/docker-syncd-brcm/start.sh:11]: not sure you should delete every file here.

**Review by lguohan — CHANGES_REQUESTED** [review]
:clock1:

File: device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/append.config.j2
Comment (gechiang) [reply device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/append.config.j2:2]: Thanks for catching it!
Have corrected it in the new change.

File: device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/append.config.j2
Comment (gechiang) [reply device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/append.config.j2:8]: Fixed.

File: device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/append.config.j2
Comment (gechiang) [reply device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/append.config.j2:20]: Moved to end.

File: device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/append.config.j2
Comment (gechiang) [reply device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-C64/append.config.j2:22]: Renamed.

File: device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-D108C8/append.config.j2
Comment (gechiang) [reply device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-D108C8/append.config.j2:9]: Moved to the end

File: device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-Q64/append.config.j2
Comment (gechiang) [reply device/arista/x86_64-arista_7260cx3_64/Arista-7260CX3-Q64/append.config.j2:11]: Renamed.

File: platform/broadcom/docker-syncd-brcm/start.sh
Comment (gechiang) [reply platform/broadcom/docker-syncd-brcm/start.sh:11]: The reason I am clearing it up is because in my manual test there were leftover generated config.bcm file and it is easier for me to debug in case things did not work and I can tell where the issue occurred if I cleaned it up every time.  
I am removing this since you are not comfortable with it.

File: platform/broadcom/docker-syncd-brcm/start.sh
Comment (gechiang) [reply platform/broadcom/docker-syncd-brcm/start.sh:24]: Renamed.

File: platform/broadcom/docker-syncd-brcm/start.sh
Comment (gechiang) [reply platform/broadcom/docker-syncd-brcm/start.sh:25]: Done

Comment (gechiang) [reply platform/broadcom/docker-syncd-brcm/start.sh:25]: Done

**Review by lguohan — COMMENTED** [review]
please wait for ying approval as well.

**Review by lguohan — DISMISSED** [review]

**Comment by gechiang** [comment]
@yxieca Appreciate if you can take a look at my changes to see if any changes required.
Thanks!

File: platform/broadcom/docker-syncd-brcm/start.sh
Comment (yxieca) [inline platform/broadcom/docker-syncd-brcm/start.sh:25]: We could re-organize this whole if structure to:

if has config.bcm.j2
   generate config.bcm and sai.profile.
elif has sai.profile.j2
   generate sai.profile
elif has sai.profile
   copy it.
endif

Comment (gechiang) [reply platform/broadcom/docker-syncd-brcm/start.sh:25]: @yxieca 
We also have to handle the case where there are no config.bcm.j2 nor sai.profile.j2.
in that case we just need to copy the sai.profile from HWSKU_DIR to the new /etc/sai.d directory.

So that is why the current start.sh changes is changed to also cover that case.
When I mentioned config.bcm.j2 and sai.profile.j2 are mutual exclusive, I should have mentioned that there is also a third case where both are absent that needs to be handled. I will update the description to address that.
Let me know if you are ok with this and if you have other concerns...

Comment (yxieca) [reply platform/broadcom/docker-syncd-brcm/start.sh:27]: The 3rd case is handled by a static sai.profile in the device folder, right? I think that if the config.bcm is auto generated, and the file name is fixed, there is no reason to leave a chance for sai.profile to be inconsistent.

If you need to handle case of generating both bcm.config and sai.profile, then you can use
if has config.bcm.j2:
   generate config.bcm and sai.profile
endif
if has sai.profile.j2
  generate sai.profile
endif
# 3rd case: use existing sai.profile in device folder.

Comment (gechiang) [reply platform/broadcom/docker-syncd-brcm/start.sh:27]: We should not use both config.bcm.j2 and sai.profile.j2.  It unnecessarily complicate things and error pron.
That is why I mentioned user should pick one or the other(mutually exclusive).
In fact, eventually we could retire the usage of sai.profile.j2 if the difference between the config files is small. and move on to use config.bcm.j2 to eliminate having multiple large config files with very small differences.  this is why Guohan asked me to make this new change.  My initial PR was using sai.profile.j2 as well and as you can see if I used that approach we will now end up with 3 large config files for "Arista-7260CX3-C64" with very small delta between them.  Two of them differ from the MMU (T0/T1) and another one for the IPinIP SOC properties if DualToR...

**Review by lguohan — APPROVED** [review]

**Pull request merged**
